JWT tutorial (#963)
mongodb/docs-realm: source/tutorial/jwt.txt

.. _jwt-tutorial:

============================================
Set up JWT Authentication with MongoDB Realm
============================================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this tutorial, you will implement :ref:`Custom JWT Authentication <config-custom-jwt>`  

> for your Realm app. You will learn how to:

Reviewer: Should this and all other instances of the word "Realm" have a source constant, such as {+backend+} for consistency with the other pages on docs-realm?

- Enable JWT Authentication,
- Use a third-party JWT provider to generate a valid token,
- Authenticate against the backend from a client application.

This tutorial should take around 30 minutes.

Prerequisites
-------------

> - Read how {+realm+} handles :ref:`Custom JWT Authentication <config-custom-jwt>`.

Author: Actually, I removed it now (based on your comment), since the tutorial itself has you do the enable/config step.

- :doc:`Create a {+app+} </get-started/create-realm-app>`. If you have completed 
  the :ref:`Task Tracker tutorial <tutorial-task-tracker-create-realm-app>`, you 
  can use that backend {+app+}.

> - Build or download one of the Task Tracker apps from the :doc:`Tutorial </tutorial>`, 

Reviewer: ```suggestion
- Build or download one of the Task Tracker apps from the :doc:`tutorial </tutorial>`, 
```

  or use one of your own.

Procedure
---------
This procedure is divided into three logical steps:

1. Create a JWT token
2. Enable and configure JWT Authentication in {+service-short+}
3. Add authentication code to your client app

Create a JWT Token
~~~~~~~~~~~~~~~~~~
For this tutorial, we will use a third-party website, `jwt.io <https://jwt.io/>`_,
to generate a JWT token. 

.. important::

   JWT tokens are credentials used to grant access to resources in 

>    your {+app+}. Care should be taken when storing them, especially when using 

Reviewer: What does "Care should be taken" really mean for a user? Who should be responsible for storing them (is it realm, some other system, a user). Perhaps linking to a wiki or guide on storing JWT tokens in development and production would be helpful?

Author: Good point. I've removed this and added an existing warning under the Signing Key, since that's the part that really matters.

   them in a production environment.

Go to `jwt.io <https://jwt.io/>`_. Scroll down until you see the **Encoded** 
and **Decoded** headers. In the **Decoded** column, you will see three sections 
that comprise a JWT token: **Header**, **Payload**, and 
**Verify Signature** areas, each of which is explained in the next sections.
 
Header
######

The header of a JWT informs consuming applications what algorithm was used to 
encode the token. The default, which we will use for this tutorial, is ``HS256``. 

.. note::

   For more information on JWT algorithms, visit `this web page 
   <https://auth0.com/blog/json-web-token-signing-algorithms-overview/>`_.
   {+service-short+} supports ``HS256`` and ``RS256``.

Payload
#######

The Payload section contains the information that we need to configure for 
our Realm-specific needs. By default, three fields are included: ``sub``, 
``name``, and ``iat``. We will use these three, plus two more fields. The following 
table explains each field and the value expected:

.. list-table::
   :header-rows: 1

   * - Field
     - Description
     - Value
   
   * - ``sub`` 

>      - The subject of the jwt. Not Used. 

Author: I was lazy. Updated.

.. important::

   The key you use must only contain ASCII letters, numbers, underscores, 
   and hyphens, and must be between 32 and 512 characters long.

.. note::

   Keep track of this secret, as you will need it in the next section when 
   we configure {+service-short+} authentication.

After pasting in your key, check the ``secret base64 encoded`` check box.

At this point, you have generated a JWT key that can be used with — and 
only with — the {+app+} you specified. Copy this key from the 
:guilabel:`Encoded` box and temporarily save it in a text document. Your JWT 
will be 3 sections of values, separated by periods, and looks something like 
the following (your token will **not** match this!):

.. code-block:: sh
   :copyable: false

   eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9. ⤶ 
   eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkNhbGViIiwiaWF0IjoxNjE3MzEzNDIwLCJle ⤶ 
   HAiOjE2NDg4NDk0MjAsImF1ZCI6ImJvaWJvaS1jeWFsOHIifQ. ⤶ 
   GaIE4e6mGxlIBNnz4FcheFBUmxd25GK-rfOfF6F85rY

Enable JWT Authentication in Your {+app+}
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you are not currently logged in to `Realm <https://realm.mongodb.com>`_ do so 

> now, and navigate to the {+app+} that you specified when generating the 

Reviewer: ```suggestion
now and navigate to the {+app+} that you specified when generating the 
```

   :guilabel:`Authentication`.

2. Select the :guilabel:`Authentication Providers` tab.

3. Near the bottom of the list of providers, select :guilabel:`Custom JWT Authentication`.

4. Use the following table to set the properties on this page:

   .. list-table::
      :header-rows: 1
      :widths: 30 70

      * - Field
        - Value
      
      * - :guilabel:`Provider Enabled`
        - Select to enable
      
      * - :guilabel:`Verification Method`
        - ``Manually specify signing keys``
      
      * - :guilabel:`Signing Algorithm`
        - ``HS256``
      
      * - :guilabel:`Signing Key (Secret Name)`
        - Provide a name that will have some meaning to you (for example, "newKey"), 
          and then click the box that appears immediately below the name that reads 
          "Create <key name>".

          In the resulting :guilabel:`Signing Key` text box, paste in the 256-bit 
          key that you used when generating the JWT in the previous section. **Do 
          not** click the "Add Singing Key" button.
      
      * - :guilabel:`Metadata Fields (Optional)`
        - Leave blank
      
      * - :guilabel:`Audience (Optional)`
        - Leave blank.

> 5. When you have set all of the values, click :guilabel:`Save`. If you are you 

Reviewer: ```suggestion
5. When you have set all of the values, click :guilabel:`Save`. If you are 
```
extra "you" here

Author: hahaha: if you are you. Well, are you, punk?

  the :ref:`Task Tracker tutorial <tutorial-task-tracker-create-realm-app>`, you 
  can use that backend {+app+}.
- You can use any {+service-short+} app for this tutorial. If you do not have 
  an test app you can use, download one of the Task Tracker apps from 
  the :doc:`tutorial </tutorial>`.

Procedure
---------
This procedure is divided into three logical steps:

1. Create a JWT token
2. Enable and configure JWT Authentication in {+service-short+}
3. Add authentication code to your client app

Create a JWT Token
~~~~~~~~~~~~~~~~~~
For this tutorial, we will use a third-party website, `jwt.io <https://jwt.io/>`_,
to generate a JWT token. 

Go to `jwt.io <https://jwt.io/>`_. Scroll down until you see the **Encoded** 
and **Decoded** headers. In the **Decoded** column, you will see three sections 
that comprise a JWT token: **Header**, **Payload**, and 
**Verify Signature** areas, each of which is explained in the next sections.
 
Header
######

The header of a JWT informs consuming applications what algorithm was used to 
encode the token. The default, which we will use for this tutorial, is ``HS256``. 

.. note::

>    For more information on JWT algorithms, visit `this web page 

Author: Yeah. I was struggling a bit with linking to a site we don't own.

          and then click the box that appears immediately below the name that reads 
          "Create <key name>".

          In the resulting :guilabel:`Signing Key` text box, paste in the 256-bit 
          key that you used when generating the JWT in the previous section. **Do 
          not** click the "Add Singing Key" button.

          .. warning::

             A :guilabel:`Signing Key` is a secret key and anyone with the
             key can issue valid user credentials for your app. Ensure that
             it's never stored in a publicly accessible location, such as a
             git repository, message board, or in your code.
      
      * - :guilabel:`Metadata Fields (Optional)`
        - Leave blank
      
      * - :guilabel:`Audience (Optional)`
        - Leave blank.

5. When you have set all of the values, click :guilabel:`Save`. If you are  
   prompted at the top of the screen to do so, review and deploy your changes. 

Update Your Client to Use JWT Authentication
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The final step in this tutorial is to update your client application to use 

> JWT authentication. The {+service-short+} SDKs make this process very simple, 

Reviewer 2: Might want to avoid saying things like "very simple" as this could make someone feel insecure. Maybe "straightforward" or "reduces the number of steps required"
```suggestion
JWT authentication. The {+service-short+} SDKs make this process straightforward, 
```

         :language: kotlin

      For more information, see :ref:`<android-login-custom-jwt>`.

   .. tab::
      :tabid: ios

      The following code snippet demonstrates implementing JWT Authentication 
      in Swift:

      .. literalinclude:: /examples/generated/code/start/Authenticate.codeblock.jwt.swift
         :language: swift
      
      For more information, see :ref:`<ios-login-custom-jwt>`

   .. tab::
      :tabid: node
      
      The following code snippet demonstrates implementing JWT Authentication 
      in Javascript:

      .. literalinclude:: /examples/generated/code/start/authenticate.codeblock.custom-jwt-login.js
         :language: javascript

      The following code snippet demonstrates implementing JWT Authentication 
      in Typescript:

      .. literalinclude:: /examples/generated/code/start/authenticate.codeblock.custom-jwt-login.ts
         :language: typescript

      For more information, see :ref:`<node-login-custom-jwt>`

   .. tab::
      :tabid: react-native
      
      The following code snippet demonstrates implementing JWT Authentication 
      in Javascript:

>       .. literalinclude:: /examples/generated/code/start/authenticate.codeblock.custom-function-login.js

Reviewer: Are these react-native examples the right links? It looks like these codeblocks are for function authentication not JWT auth.

Author: ooops! Good catch!

         :language: javascript

>       The following code snippet demonstrates implementing JWT Authentication 

Reviewer: Can you remove these typescript examples? They're the same as the JS examples right above. The unit test file authenticate.ts needs to be updated at some point (which I can do as part of my unit test ticket)
